engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java

```diff
@@ -2424,4 +2435,103 @@ public int hashCode()
     return result;
   }
 
+  /**
+   * Return list of operators in topological sorted order. The Delay operators
+   * are considered as downstream operator.
+   * @return
+   */
+  @VisibleForTesting
+  protected List<OperatorMeta> getSortedOperators()
+  {
+    List<OperatorMeta> sortedOperators = new ArrayList<>(operators.size());
+    Stack<OperatorMeta> pendingNodes = new Stack<>();
+    Set<OperatorMeta> visited = new HashSet<>();
+
+    for (OperatorMeta n : this.getAllOperators()) {
+      pendingNodes.push(n);
+    }
+
+    while (!pendingNodes.isEmpty()) {
+      OperatorMeta n = pendingNodes.pop();
+
+      if (visited.contains(n)) {
+        continue;
+      }
+
+      boolean upstreamVisited = true;
+      for (Map.Entry<InputPortMeta, StreamMeta> entry : n.getInputStreams().entrySet()) {
+        StreamMeta s = entry.getValue();
+        boolean delay = entry.getKey().getValue(LogicalPlan.IS_CONNECTED_TO_DELAY_OPERATOR);
+        // skip delay sources since it's going to be handled as downstream
+        if (!delay && s.getSource() != null && !visited.contains(s.getSource().getOperatorMeta())) {
+          pendingNodes.push(n);
+          pendingNodes.push(s.getSource().getOperatorMeta());
+          upstreamVisited = false;
+          break;
+        }
+      }
+
+      if (upstreamVisited) {
+        visited.add(n);
+        sortedOperators.add(n);
+      }
+    }
+
+    return sortedOperators;
+  }
+
+  /**
+   * Get list of stream from root operator to downstreams. The stream connecting
+   * from delay operators are considered as downstream streams.
+   * @return
+   */
+  @VisibleForTesting
+  protected List<StreamMeta> getSortedStreams()
```

Comment:
Call this getStreamsTopographicalOrder() and make it available in the interface, see lifecycle comment below.